server/api/authentication/authenticator.go

```diff
@@ -0,0 +1,35 @@
+package authentication
+
+import (
+	"crypto/hmac"
+	"crypto/sha512"
+	"encoding/hex"
+	"github.com/tobocop/go-teleport-directory-browser/api/user"
+	"golang.org/x/crypto/bcrypt"
+)
+
+type Authenticator interface {
+	Authenticate(username string, password string) (bool, error)
+}
+
+// I'd provide this via a credential manager (aws secrets, credhub, etc) or via an environment variable that was populated through more secure means
+var hmacKey = []byte("8cf7a749-bf77-42ad-abc4-7cf110872bc4")
+
+type staticCredentialsAuthenticator struct {
+	UserStore user.Store
+}
+
+func (sa *staticCredentialsAuthenticator) Authenticate(username string, password string) (bool, error) {
+	mac := hmac.New(sha512.New, hmacKey)
+	mac.Write([]byte(password))
+	hash := hex.EncodeToString(mac.Sum(nil))
+
+	hashed, err := bcrypt.GenerateFromPassword([]byte(hash), 10)
+	if err != nil {
+		return false, err
+	}
```

Comment:
Why are you passing a `hmac(password)` to bcrypt?

Reply:
Because the stored password it will be compared against is hashed as well. The idea was to prevent storage of plain-text passwords even though it's hard-coded. Maybe I've miss-understood how bcrypt works though.

Reply:
I'm still not sure I understand the need for the hmac?

The output of `bcrypt.GenerateFromPassword` is a string that looks like the following which embeds the salt + hash (and other information) in it. You can pre-generate this and just check it into source.

```
$2a$10$J2dMyKwK4188qqjTnk/vkurcdEUMRt/Bq0zoa6isxdsxHWKIPFoyu
```

Then all you have to do is call `bcrypt.CompareHashAndPassword` with the above and the passed in password.

Using bcrypt alone this way, I don't see how a plaintext password would be stored? Maybe I am missing something?

Reply:
So, that's what I started with, and the comparison failed every time. I looked into it and the problem is the golang bcrypt doesn't allow you to set an explicit salt, so every time the hash is generated is uses a unique salt. Its comparison function will take a plain text value and generate a hash based on the salt from the hashed password (first arg). So basically, you won't get the same value every time for the same password. I made a go playground to illustrate what's going on:
https://play.golang.org/p/fRs_u1ahpfq

I think in a real project I'd either look for a different bcrypt implementation in which we could explicitly control the salt or I'd look for a nother suitable algorithm.

Reply:
Disregard above, I made a mistake with my compare implementation in that it didn't need to re-generate the hash from the user provided password. Code should be good now.